ledger/narwhal/compact-batch-certificate/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+use narwhal_transmission_id::TransmissionType;
+
+use core::hash::{Hash, Hasher};
+use indexmap::{IndexMap, IndexSet};
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct CompactBatchCertificate<N: Network> {
+    /// The certificate ID.
+    certificate_id: Field<N>,
+    /// The compact batch header.
+    compact_batch_header: CompactBatchHeader<N>,
+    /// The `(signature, timestamp)` pairs for the batch ID from the committee.
+    signatures: IndexMap<Signature<N>, i64>,
+}
+
+impl<N: Network> CompactBatchCertificate<N> {
+    /// Initializes a new compact batch certificate.
+    pub fn new(
+        compact_batch_certificate: CompactBatchHeader<N>,
+        signatures: IndexMap<Signature<N>, i64>,
+    ) -> Result<Self> {
+        // Compute the certificate ID.
+        let certificate_id = Self::compute_certificate_id(compact_batch_certificate.batch_id(), &signatures)?;
+        // Return the compact batch certificate.
+        Self::from(certificate_id, compact_batch_certificate, signatures)
+    }
+
+    /// Initializes a new compact batch certificate.
+    pub fn from(
+        certificate_id: Field<N>,
+        compact_batch_header: CompactBatchHeader<N>,
+        signatures: IndexMap<Signature<N>, i64>,
+    ) -> Result<Self> {
+        // Compute the certificate ID.
+        if certificate_id != Self::compute_certificate_id(compact_batch_header.batch_id(), &signatures)? {
+            bail!("Invalid batch certificate ID")
+        }
+        // Verify the signatures are valid.
+        for (signature, timestamp) in &signatures {
+            let preimage = [compact_batch_header.batch_id(), Field::from_u64(*timestamp as u64)];
+            if !signature.verify(&signature.to_address(), &preimage) {
+                bail!("Invalid compact batch certificate signature")
+            }
+        }
+        // Return the compact batch certificate.
+        Self::from_unchecked(certificate_id, compact_batch_header, signatures)
+    }
+
+    /// Initializes a new compact batch certificate.
+    pub fn from_unchecked(
+        certificate_id: Field<N>,
+        compact_batch_header: CompactBatchHeader<N>,
+        signatures: IndexMap<Signature<N>, i64>,
+    ) -> Result<Self> {
+        // Ensure the signatures are not empty.
+        ensure!(!signatures.is_empty(), "Batch certificate must contain signatures");
+        // Return the compact batch certificate.
+        Ok(Self { certificate_id, compact_batch_header, signatures })
+    }
+
+    /// Initializes a new compact batch certificate from a batch certificate.
+    pub fn from_batch_certificate(batch_certificate: &BatchCertificate<N>) -> Result<Self> {
+        // Construct the compact batch header.
+        let compact_batch_header = CompactBatchHeader::from_batch_header(batch_certificate.batch_header())?;
+
+        // Construct the signatures.
+        let signatures = batch_certificate
+            .signatures()
+            .zip_eq(batch_certificate.timestamps())
```

Comment:
we might want to first compare lengths and return an error, `zip_eq` will panic on different lengths after all